content/courses/advanced-ada/chapters/numerics.rst

```diff
@@ -1061,26 +1061,115 @@ Let's see some examples:
 
-        procedure Show_Copy_Sign_Leading_Part_Machine is
-        begin
-           --  NOTE: no clear usage for 'Machine!!
-           Put_Line (Float'(1.000015)'Image);
-           Put_Line (Float'Machine (1.000015)'Image);
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Float_Value is
+       package F_IO is new Ada.Text_IO.Float_IO (Float);
+
+       V : Float;
+    begin
+       F_IO.Default_Fore := 3;
+       F_IO.Default_Aft  := 1;
+       F_IO.Default_Exp  := 0;
+
+       V := 1.0E+15;
+       Put ("1.0E+15 = ");
+       F_IO.Put (Item => V);
+       New_Line;
+
+    end Show_Float_Value;
+
+If we run this example on a typical PC, we see that the expected value
+1000000000000000.0 was displayed as 999999986991000.0.
+This is because 1.0 x 10\ :sup:`15` isn't
+directly representable on this machine, so it has to be modified to a value that
+is actually representable (on the machine). This modification can be made
+visible with the :ada:`'Machine (X)` attribute, which returns a version of
+:ada:`X` that is representable on the target machine. This modification is
+performed by rounding or truncating :ada:`X` to either one of the adjacent
+machine numbers for the specific floating-point type of :ada:`X`. (Of course, if
+the real value of :ada:`X` is representable on the target machine, no
+modification is performed.)
+
+In fact, we could rewrite the :ada:`V := 1.0E+15` assignment of the code example
+as :ada:`V := Float'Machine (1.0E+15)`, as we're never assigning a real value
+directly to a floating-pointing variable, but instead first converting it to a
+version of the real value that is representable on the target machine. In this
+case, 999999986991000.0 is a representable version of the real value
+1.0 x 10\ :sup:`15`. Of course, writing :ada:`V := 1.0E+15` or
+:ada:`V := Float'Machine (1.0E+15)` doesn't make any difference to the actual
+value that is assigned to :ada:`V`, as the conversion to a representable value
+happens automatically during the assignment to :ada:`V`.
+
+There are, however, instances where using the :ada:`'Machine` attribute does
+make a difference in the result. For example, let's say we want to calculate
+the difference between the original real value in our example
+(1.0 x 10\ :sup:`15`) and the actual value that is assigned to :ada:`V`. We can
+do this by using the :ada:`'Machine` attribute in the calculation:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Numerics.Machine_Attribute
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Machine_Attribute is
+       package F_IO is new Ada.Text_IO.Float_IO (Float);
+
+       V : Float;
+    begin
+       F_IO.Default_Fore := 3;
+       F_IO.Default_Aft  := 1;
+       F_IO.Default_Exp  := 0;
+
+       Put_Line ("Original value: 1000000000000000.0");
+
+       V := 1.0E+15;
+       Put ("Machine value:  ");
+       F_IO.Put (Item => V);
+       New_Line;
+
+       V := 1.0E+15 - Float'Machine (1.0E+15);
```

Comment:
It's good that you've asked Tuck to review this section. I would have bet that V was assigned zero here!